### ISSUE TITLE
windows python crash after exiting spock

### ISSUE THREAD

**Issue by dschick** [sardana-org/sardana#724]
Hej,

when I exit spock under windows 10, python always crashes with the following error message in the spock terminal

```
MainThread     INFO     2018-03-27 12:31:28,565 TaurusRootLogger:
*********************
< Deprecation Counts (525):
        407 * "_get_value is deprecated since 4.0. Use .rvalue instead"
        67 * "getHWObj is deprecated since 4.0. Use getDeviceProxy() instead"
        25 * "getState is deprecated since 4.0. Use .stateObj.read().rvalue [Tango] or .state [agnostic] instead"
        25 * "getSWState is deprecated since 4.0. Use state instead"
        1 * "getStateObj is deprecated since 4.0. Use .stateObj [Tango] or .factory.getAttribute(state_full_name) [agnostic] instead" >
Assertion failed: Successful WSASTARTUP not yet performed (..\..\..\src\signaler.cpp:377)
```

**Comment by cpascual** [comment]
This part is normal (just debugging messages from the compatibility layer for taurus3)

```
< Deprecation Counts (525):
        407 * "_get_value is deprecated since 4.0. Use .rvalue instead"
        67 * "getHWObj is deprecated since 4.0. Use getDeviceProxy() instead"
        25 * "getState is deprecated since 4.0. Use .stateObj.read().rvalue [Tango] or .state [agnostic] instead"
        25 * "getSWState is deprecated since 4.0. Use state instead"
        1 * "getStateObj is deprecated since 4.0. Use .stateObj [Tango] or .factory.getAttribute(state_full_name) [agnostic] instead" >
``` 

But regarding the last line  (`Assertion failed: (...)`) I do not know where it comes from or if it is really problematic.

**Comment by dschick** [comment]
despite of these messages in the terminal there is also a popup claiming that python has crashed and needs to be closed.
Nothing serious but in order to convince my colleagues to use sardana it would be helpful to get rid of most of the dramatic error messages ;)

thanks for the quick reply

**Comment by cmft** [comment]
We had seen this kinds of problems in some Windows setups. In order to try to reproduce these errors can you report us which version (python, tango, qt, pytango, taurus and sardana ) and how did you install them?.

**Comment by dschick** [comment]
hi @cmft ,
I used WinPython 2.7.13.1 from [here](https://sourceforge.net/projects/winpython/files/WinPython_2.7/2.7.13.1/)

- tango 9.2.1 installed from executable (https://pypi.python.org/pypi/PyTango/9.2.1#downloads)
- taurus 4.3.1 installed via pip; I used the compiled packages from [here](https://www.lfd.uci.edu/~gohlke/pythonlibs/#pyqt4) for the PyQt-dependencies of taurus
- Sardana 2.4.0 installed via pip
- PyQt 4.1

Btw: I got the same error with the `Assertion failed` also within Spock when I start an expconf with a misconfigured measurement group (i.e. I deleted a counter and its controller via jive without prior deleting it from the measurement group). Although there was again the python crash pop-up I could continiue using spock then

**Comment by cpascual** [comment]
So it looks like the Assertion failed message is indeed related to the crash. And it is likely related to some problem during the termination of a Qt application 
Note that you wrote "PyQt 4.1" that is *ancient* and certainly not supported... maybe you mean 4.11?.

If it is not much problem, would it be possible for you to try a installation from scratch on a clean environment using the following recipe:

**Comment by cpascual** [comment]
Alternatively, here is the recipe we use in our Vagrant test images to install the taurus and sardana dependencies on windows:

```
      C:\\vagrant\\Python(x,y)-2.7.10.0.exe
      C:\\Python27\\python.exe -m pip install --upgrade pip
      msiexec /quiet /i C:\\vagrant\\pytango-9.2.1.Win32-py2.7.msi | out-null
      C:\\Python27\\Scripts\\pip.exe install Cython==0.27.3
      C:\\Python27\\Scripts\\pip.exe install --upgrade numpy==1.14.1
      C:\\Python27\\Scripts\\pip.exe install qtpy==1.3.1
      C:\\Python27\\Scripts\\pip.exe install Pillow==5.0.0
      C:\\Python27\\Scripts\\pip.exe install PythonQwt==0.5.5
      C:\\Python27\\Scripts\\pip.exe install --upgrade guidata==1.7.6
      C:\\Python27\\Scripts\\pip.exe install --upgrade --no-deps scipy==1.0.0
      C:\\Python27\\Scripts\\pip.exe install --upgrade spyder==3.2.7
      C:\\Python27\\Scripts\\pip.exe install --upgrade --no-deps guiqwt==3.0.3
      C:\\Python27\\Scripts\\pip.exe install PyMca5==5.2.2

```

**Comment by dschick** [comment]
Hej @cpascual ,

thanks for the recipe! btw you should publish this in the sardana documentation.

However, I got exactly the same errors as with the WinPython distribution.
With WinPython I installed indeed PyQt 4.11.4 not 4.1

Also the other bugs which where reportet by @mrosanes in #717 & #716 as well as #91 are independent of python distribution for my system. 

I can try using Win7 instead of Win10 next week.
Can I report any logs here for further debugging?

**Comment by reszelaz** [comment]
> thanks for the recipe! btw you should publish this in the sardana documentation.

I totally agree, we should have it somewhere public. But for the moment I don't have any clear idea where to put it. Just to open the brainstorming:
* DirectlyIn the documentation
* In the release manual test markdown document
* As powershell script in the CI directory

Actually we were thinking in publishing the whole Vagrant configuration file as soon as we clarify the Windows licensing issues. Let's see if we can push it soon...

> I can try using Win7 instead of Win10 next week.

If it is just to confirm this bug, I think it is not worth. When working on #717 (on a Win7 Vagrant box with the above installation instructions) I have observed that it is not possible to exit from spock. Making the whole CMD window useless. There is no crash, neither the AssertionError message, but it simply hangs. So certainly there is something wrong.

> Can I report any logs here for further debugging?

If you advance with this topic, please don't hesitate to post any comments, log files, etc.

**Comment by dschick** [comment]
I, as an early user, would appriciate the windows recipe directly in the documentation, where are already some of the required steps mentioned. For me that would be the first location to look for it.

For Windows I would think of an extended python distribution where Sardana is already installed with all required dependencies. But I guess that is not easily possible because of the PyQt and and tango dependencies, right?

For now I am not able to generate any further error logs than the one I posted above. But as I said, I sometimes encounter the same/a similar error when I exit the expconf editor.
Can I start spock via running it with a python command from the CMD? Then I might use the -m trace option or maybe I get some additional errors in the CMD window.

**Comment by cpascual** [comment]
Hi, I just made a new issue (#727) to move the open discussion on the installation recipe there so that we can use this one for focusing on the original crash problem.

**Comment by dschick** [comment]
@reszelaz when I change from sardana 2.4.0 to 2.4.1-alpha the cmd also stucks when exiting without any error message.

**Comment by reszelaz** [comment]
While testing Jul20 release (3.0.2) the spock still hangs the whole CMD window at exiting.

**Comment by reszelaz** [comment]
Note that here we may be hitting https://github.com/tango-controls/pytango/issues/390.

**Comment by reszelaz** [comment]
We could try this hint: https://github.com/tango-controls/pytango/issues/390#issuecomment-746995003 (the part of the client).

**Comment by reszelaz** [comment]
Fixed by upgrading to Taurus 4.7.1.
